api/envoy/config/route/v3/route_components.proto

```diff
@@ -1564,3 +1567,74 @@ message QueryParameterMatcher {
     bool present_match = 6;
   }
 }
+
+// HTTP Internal Redirect :ref:`architecture overview <arch_overview_internal_redirects>`.
+// [#next-free-field: 6]
+message InternalRedirectPolicy {
+  // Configures :ref:`internal redirect <arch_overview_internal_redirects>` behavior.
+  enum InternalRedirectAction {
+    // Redirect responses are proxied to downstream without triggering Internal Redirect.
+    // This is the default value.
+    PASS_THROUGH_INTERNAL_REDIRECT = 0;
+
+    // Redirect responses (specified via :ref:`redirect_response_codes
+    // <envoy_api_field_config.route.v3.InternalRedirectPolicy.redirect_response_codes>`),
+    // are followed by Envoy to the new URL. The redirect responses are not visible to
+    // the downstream unless the redirect following is rejected by other criteria.
+    HANDLE_INTERNAL_REDIRECT = 1;
+  }
+
+  // HTTP scheme combinations that are allowed to be handled as internal
+  // redirect. Downstream scheme comes first and redirect target URL scheme
+  // follows.
+  enum DownstreamAndRedirectTargetSchemePair {
+    HTTP_HTTP = 0;
+    HTTP_HTTPS = 1;
+    HTTPS_HTTP = 2;
+    HTTPS_HTTPS = 3;
+  }
+
+  message Predicate {
+    string name = 1 [(validate.rules).string = {min_bytes: 1}];
+
+    google.protobuf.Any typed_config = 2;
+  }
+
+  InternalRedirectAction internal_redirect_action = 1;
+
+  // An internal redirect is handled, iff the number of previous internal redirects that a
+  // downstream request has encountered is lower than this value, and
+  // :ref:`internal_redirect_action
+  // <envoy_api_field_config.route.v3.InternalRedirectPolicy.internal_redirect_action>`
+  // is set to
+  // :ref:`HANDLE_INTERNAL_REDIRECT
+  // <envoy_api_enum_value_config.route.v3.InternalRedirectPolicy.InternalRedirectAction.HANDLE_INTERNAL_REDIRECT>`
+  // In the case where a downstream request is bounced among multiple routes by internal redirect,
+  // the first route that hits this threshold, or has
+  // :ref:`internal_redirect_action
+  // <envoy_api_field_config.route.v3.InternalRedirectPolicy.internal_redirect_action>`
+  // set to
+  // :ref:`PASS_THROUGH_INTERNAL_REDIRECT
+  // <envoy_api_enum_value_config.route.v3.InternalRedirectPolicy.InternalRedirectAction.PASS_THROUGH_INTERNAL_REDIRECT>`
+  // will pass the redirect back to downstream.
+  //
+  // If not specified, at most one redirect will be followed.
+  google.protobuf.UInt32Value max_internal_redirects = 2;
+
+  // Defines what upstream response codes are allowed to trigger internal redirect. This is ignored unless
+  // :ref:`internal_redirect_action
+  // <envoy_api_field_config.route.v3.InternalRedirectPolicy.internal_redirect_action>`
+  // is set to
+  // :ref:`HANDLE_INTERNAL_REDIRECT
+  // <envoy_api_enum_value_config.route.v3.InternalRedirectPolicy.InternalRedirectAction.HANDLE_INTERNAL_REDIRECT>`
+  repeated uint32 redirect_response_codes = 3;
```

Comment:
Do we want to bound this?  2xx redirects feel sketchy and 1xx seem terrifying :-P

Reply:
Yep. I ideally want to use repeated.items.uint32 {in = [301, 302, 303, 307, 308]}, but unfortunately, pgv does not generate proper "items.uint32.in" validation code for C++, so I switched to lte and gte. I saw there is a PR in pgv that may fix this problem. I'll tight it down when it's in.

https://github.com/envoyproxy/protoc-gen-validate/pull/330

Reply:
In general, more strict PGV annotations are breaking API changes, but since in this case there is a documented contract that these restrictions exist, I agree that we can do this later.